src/spi/bus.rs

```diff
@@ -0,0 +1,369 @@
+        self.spi.ie.write(|w| w.rxwm().clear_bit())
+    }
+
+    /// Set AUTO CS mode to per-word operation, unless CSMODE is set to OFF
+    pub fn cs_mode_word(&mut self) {
+        if !self.spi.csmode.read().mode().is_off() {
+            self.spi.csmode.write(|w| w.mode().auto());
+        }
+    }
+
+    /// Set HOLD CS mode to per-frame operation, unless CSMODE is set to OFF
+    pub fn cs_mode_frame(&mut self) {
+        if !self.spi.csmode.read().mode().is_off() {
+            self.spi.csmode.write(|w| w.mode().hold());
+        }
+    }
+
+    /// Finishes transfer by deasserting CS (only for hardware-controlled CS)
+    pub fn end_transfer(&mut self) {
+        self.cs_mode_word()
+    }
+
+    /// Releases the SPI peripheral and associated pins
+    pub fn free(self) -> (SPI, PINS) {
+        (self.spi, self.pins)
+    }
+
+    /// Configure the SPI bus for given mode, frequency and CS index
+    pub(crate) fn configure(
+        &mut self,
+        mode: Mode,
+        clock_divisor: u32,
+        cs_mode: e310x::qspi0::csmode::MODE_A,
+        cs_index: Option<u32>,
+    ) where
+        PINS: Pins<SPI>,
+    {
+        assert!(clock_divisor <= 0xfff);
+        self.spi
+            .sckdiv
+            .write(|w| unsafe { w.div().bits(clock_divisor as u16) });
+
+        if let Some(index) = cs_index {
+            self.spi.csid.write(|w| unsafe { w.bits(index) });
+        }
+        self.spi.csmode.write(|w| w.mode().variant(cs_mode));
+
+        // Set CS pin polarity to high
+        self.spi.csdef.reset();
+
+        // Set SPI mode
+        let phase = mode.phase == Phase::CaptureOnSecondTransition;
+        let polarity = mode.polarity == Polarity::IdleHigh;
+        self.spi
+            .sckmode
+            .write(|w| w.pha().bit(phase).pol().bit(polarity));
+
+        self.spi.fmt.write(|w| unsafe {
+            w.proto().single();
+            w.endian().big(); // Transmit most-significant bit (MSB) first
+            w.dir().rx();
+            w.len().bits(8)
+        });
+
+        // Set watermark levels
+        self.spi.txmark.write(|w| unsafe { w.txmark().bits(1) });
+        self.spi.rxmark.write(|w| unsafe { w.rxmark().bits(0) });
+
+        // set delays to their reset values
+        self.spi.delay0.write(|w| unsafe {
+            w.cssck().bits(1); // no delay on CS assert
+            w.sckcs().bits(1) // 1 SCK cycle delay on CS deassert
+        });
+        self.spi.delay1.write(|w| unsafe {
+            w.intercs().bits(1); // min 1 SCK cycle between CS re-assets
+            w.interxfr().bits(0) // no intra-frame delay
+        });
+    }
+
+    /// Force CS assert by switching to HOLD when bus is unshared only
+    fn assert_cs(&mut self) {
+        if !self.shared {
+            self.cs_mode_frame()
+        }
+    }
+
+    /// Force CS deassert by switching to AUTO when bus is unshared only
+    fn deassert_cs(&mut self) {
+        if !self.shared {
+            self.cs_mode_word()
+        }
+    }
```

Comment:
I think "Force CS (de)assert" is not correct here as CS changes before the first byte and after the last byte, and not always right after the mode change. "Prepare CS (de)assert" is probably better.

Reply:
will do